File: Progetto-03/Info e Slide/seiot1920/java/src/seiot/modulo_lab_4_3/MsgService.java
package seiot.modulo_lab_4_3;

import seiot.modulo_lab_2_2.msg.CommChannel;
import seiot.modulo_lab_4_2.ExtendedSerialCommChannel;

public class MsgService extends seiot.modulo_lab_3_2.common.Observable {

	private CommChannel channel;
	private String port;
	private int rate;
	
	public MsgService(String port, int rate){
		this.port = port;
		this.rate = rate;
	}
	
	void init(){
		try {
			channel = new ExtendedSerialCommChannel(port, rate);	
			// channel = new SerialCommChannel(port, rate);	
			System.out.println("Waiting Arduino for rebooting...");		
			Thread.sleep(4000);
			System.out.println("Ready.");		
		} catch (Exception e) {
			e.printStackTrace();
		}
		
		new Thread(() -> {
			while (true) {
				try {
					String msg = channel.receiveMsg();
					System.out.println("received "+msg);
					this.notifyEvent(new MsgEvent(msg));
				} catch (Exception ex) {
					ex.printStackTrace();
				}
			}
		}).start();
	}
	
	public void sendMsg(String msg) {
		channel.sendMsg(msg);
		System.out.println("sent "+msg);
	}
	
}

File: Progetto-03/Info e Slide/seiot1920/arduino/modulo-lab-4.1/bt_button_led/MsgService.cpp
#include "Arduino.h"
#include "MsgService.h"


MsgService::MsgService(int rxPin, int txPin){
  channel = new SoftwareSerial(rxPin, txPin);
}

void MsgService::init(){
  content.reserve(256);
  channel->begin(9600);
}

bool MsgService::sendMsg(Msg msg){
  channel->println(msg.getContent());  
}

bool MsgService::isMsgAvailable(){
  return channel->available();
}

Msg* MsgService::receiveMsg(){
  while (channel->available()) {
    char ch = (char) channel->read();
    if (ch == '\n'){
      Msg* msg = new Msg(content);  
      content = "";
      return msg;    
    } else {
      content += ch;      
    }
  }
  return NULL;  
}
